pkg/policy/repository.go

```diff
@@ -351,51 +353,121 @@ nextLabel:
+	_, rev := p.AddListLocked(newList)
+	return rev, map[uint16]struct{}{}, nil
+}
+
+// NewIDSet returns a new instance of an IDSet.
+func NewIDSet() *IDSet {
+	return &IDSet{
+		IDs: map[uint16]struct{}{},
+	}
+}
+
+type IdentityConsumer interface {
+	GetID16() uint16
+	GetSecurityIdentity() *identity.Identity
+	PolicyRevisionBumpEvent(rev uint64)
 }
 
 // AddListLocked inserts a rule into the policy repository with the repository already locked
 // Expects that the entire rule list has already been sanitized.
-func (p *Repository) AddListLocked(rules api.Rules) uint64 {
-	newList := make([]*rule, len(rules))
+func (p *Repository) AddListLocked(rules api.Rules) (ruleSlice, uint64) {
+
+	newList := make(ruleSlice, len(rules))
 	for i := range rules {
-		newList[i] = &rule{Rule: *rules[i]}
+		newRule := &rule{
+			Rule:     *rules[i],
+			metadata: newRuleMetadata(),
+		}
+		newList[i] = newRule
 	}
+
 	p.rules = append(p.rules, newList...)
 	p.revision++
 	metrics.PolicyCount.Add(float64(len(newList)))
 	metrics.PolicyRevision.Inc()
 
-	return p.revision
+	return newList, p.revision
+}
+
+// UpdateLocalConsumers updates the cache within each rule in the given repository
+// which specifies whether said rule selects said identity. Returns a wait group
+// which can be used to wait until all rules have had said caches updated.
+func (p *Repository) UpdateLocalConsumers(identifiers []IdentityConsumer) *sync.WaitGroup {
+	var policySelectionWG sync.WaitGroup
+	for _, identityConsumer := range identifiers {
+		policySelectionWG.Add(1)
+		go p.rules.analyzeWhetherRulesSelectEndpoint(identityConsumer, NewIDSet(), &policySelectionWG)
+	}
+	return &policySelectionWG
 }
 
-// AddList inserts a rule into the policy repository.
-func (p *Repository) AddList(rules api.Rules) uint64 {
+// RemoveIdentifierFromRuleCaches removes identifier from the processedConsumers
+// and localRuleConsumers sets in each rule within the repository. Returns a
+// sync.WaitGroup which can be waited on once all rules have been processed in
+// relation to the identifier.
+func (p *Repository) RemoveIdentifierFromRuleCaches(identifier uint16) *sync.WaitGroup {
+	p.Mutex.RLock()
+	defer p.Mutex.RUnlock()
+	var wg sync.WaitGroup
+	wg.Add(len(p.rules))
+	for _, r := range p.rules {
+		go func(rr *rule, wgg *sync.WaitGroup) {
+			rr.metadata.deleteID(identifier)
+			wgg.Done()
+		}(r, &wg)
+	}
+	return &wg
+}
+
+// AddList inserts a rule into the policy repository. It is used for
+// unit-testing purposes only.
+func (p *Repository) AddList(rules api.Rules) (ruleSlice, uint64) {
 	p.Mutex.Lock()
 	defer p.Mutex.Unlock()
 	return p.AddListLocked(rules)
 }
 
+// IDSet is a wrapper type around a set of unsigned 16-bit integers, with
+// a mutex for protecting access.
+type IDSet struct {
+	Mutex lock.RWMutex
+	IDs   map[uint16]struct{}
+}
+
+func (r ruleSlice) UpdateEndpointsAffectedByRules(localConsumers []IdentityConsumer, consumersToUpdate *IDSet, policySelectionWG *sync.WaitGroup) {
```

Comment:
nit: `UpdateEndpointCache()`?

Reply:
Wait, is this trying to update the cache for these or is it trying to determine whether the rules select the endpoint? I think in different callers of this function we might need only the former, or we might need both.

Reply:
It does both things - the cache within the rules is updated whenever it is analyzed in relation to a given endpoint ID / security identity pairing.